---
string statement
string[] predicates
string[] functions